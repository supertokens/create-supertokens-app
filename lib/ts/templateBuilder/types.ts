// types/config.ts
export type ConfigType =
    | "all_auth"
    | "emailpassword"
    | "multifactorauth"
    | "multitenancy"
    | "passwordless"
    | "thirdparty"
    | "thirdpartyemailpassword"
    | "thirdpartypasswordless"
    | "webauthn";

export type IndividualRecipe =
    | "emailPassword"
    | "thirdParty"
    | "passwordless"
    | "session"
    | "dashboard"
    | "userRoles"
    | "multiFactorAuth"
    | "accountLinking"
    | "emailVerification"
    | "totp"
    | "multitenancy"
    | "webauthn";

export type FrontendFramework = "react" | "vue" | "angular" | "solid";

export type Language = "ts" | "py" | "go";

export type AppInfo = {
    appName: string;
    apiDomain: string;
    websiteDomain: string;
    apiBasePath: string;
    websiteBasePath: string;
    connectionURI: string;
};

export type OAuthProvider = {
    id: string;
    clientId: string;
    clientSecret: string;
    additionalConfig?: Record<string, any>;
};
