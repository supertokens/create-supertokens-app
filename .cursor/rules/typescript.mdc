---
description: 
globs: 
alwaysApply: false
---
## TypeScript

| factor          | ts imports                                                             | ts inits                                                                                                                                                                    |
| --------------- | ---------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `emailpassword` | import EmailPassword from "supertokens-node/recipe/emailpassword";<br> | EmailPassword.init()                                                                                                                                                        |
| `thirdparty`    | import ThirdParty from "supertokens-node/recipe/thirdparty";           | ThirdParty.init({<br>            signInAndUpFeature: {<br>                providers: [<br>    // per-provider config<br>                ],<br>            },<br>        }), |
| `link_email`    | import Passwordless from "supertokens-node/recipe/passwordless";       | Passwordless.init({<br>            contactMethod: "EMAIL",<br>            flowType: MAGIC_LINK",<br>        }),                                                             |
| `link_phone`    | import Passwordless from "supertokens-node/recipe/passwordless";       | Passwordless.init({<br>            contactMethod: "PHONE",<br>            flowType: "MAGIC_LINK",<br>        }),                                                            |
| `otp_phone`     | import Passwordless from "supertokens-node/recipe/passwordless";       | Passwordless.init({<br>            contactMethod: "PHONE",<br>            flowType: "USER_INPUT_CODE",<br>        }),                                                       |
| `otp_email`     | import Passwordless from "supertokens-node/recipe/passwordless";       | Passwordless.init({<br>            contactMethod: "EMAIL",<br>            flowType: "USER_INPUT_CODE",<br>        }),                                                       |
| `totp`          | import totp from "supertokens-node/recipe/totp"                        | totp.init(),                                                                                                                                                                |

**Notes**: 
- Session, Dashboard, UserRoles also imported for all factors, in addition to their own config.
- if both link_email and link_phone are present as factors, the contactMethod in init becomes "EMAIL_OR_PASSWORD".
- if both otp_email and otp_phone are present as factors, the flowType in init becomes "USER_INPUT_CODE_AND_MAGIC_LINK".
- If a second factor is present, always include the MultiFactorAuth recipe.
- For any MultiFactor recipes, also include AccountLinking and EmailVerification.
- Second factors specific behavior always go in the override function of the MultiFactorAuth recipe!


### Examples

#### All auth

```ts
import EmailPassword from "supertokens-node/recipe/emailpassword";
import ThirdParty from "supertokens-node/recipe/thirdparty";
import Passwordless from "supertokens-node/recipe/passwordless";
import Session from "supertokens-node/recipe/session";
import { TypeInput } from "supertokens-node/types";
import Dashboard from "supertokens-node/recipe/dashboard";
import UserRoles from "supertokens-node/recipe/userroles";

export function getApiDomain() {
    const apiPort = process.env.VITE_APP_API_PORT || 3001;
    const apiUrl = process.env.VITE_APP_API_URL || `http://localhost:${apiPort}`;
    return apiUrl;
}

export function getWebsiteDomain() {
    const websitePort = process.env.VITE_APP_WEBSITE_PORT || 3000;
    const websiteUrl = process.env.VITE_APP_WEBSITE_URL || `http://localhost:${websitePort}`;
    return websiteUrl;
}

export const SuperTokensConfig: TypeInput = {
    supertokens: {
        // this is the location of the SuperTokens core.
        connectionURI: "https://try.supertokens.com",
    },
    appInfo: {
        appName: "SuperTokens Demo App",
        apiDomain: getApiDomain(),
        websiteDomain: getWebsiteDomain(),
    },
    // recipeList contains all the modules that you want to
    // use from SuperTokens. See the full list here: https://supertokens.com/docs/guides
    recipeList: [
        EmailPassword.init(),
        ThirdParty.init({
            signInAndUpFeature: {
                providers: [
                    // We have provided you with development keys which you can use for testing.
                    // IMPORTANT: Please replace them with your own OAuth keys for production use.
                    {
                        config: {
                            thirdPartyId: "google",
                            clients: [
                                {
                                    clientId:
                                        "1060725074195-kmeum4crr01uirfl2op9kd5acmi9jutn.apps.googleusercontent.com",
                                    clientSecret: "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",
                                },
                            ],
                        },
                    },
                    {
                        config: {
                            thirdPartyId: "github",
                            clients: [
                                {
                                    clientId: "467101b197249757c71f",
                                    clientSecret: "e97051221f4b6426e8fe8d51486396703012f5bd",
                                },
                            ],
                        },
                    },
                    {
                        config: {
                            thirdPartyId: "apple",
                            clients: [
                                {
                                    clientId: "4398792-io.supertokens.example.service",
                                    additionalConfig: {
                                        keyId: "7M48Y4RYDL",
                                        privateKey:
                                            "-----BEGIN PRIVATE KEY-----\nMIGTAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBHkwdwIBAQQgu8gXs+XYkqXD6Ala9Sf/iJXzhbwcoG5dMh1OonpdJUmgCgYIKoZIzj0DAQehRANCAASfrvlFbFCYqn3I2zeknYXLwtH30JuOKestDbSfZYxZNMqhF/OzdZFTV0zc5u5s3eN+oCWbnvl0hM+9IW0UlkdA\n-----END PRIVATE KEY-----",
                                        teamId: "YWQCXGJRJL",
                                    },
                                },
                            ],
                        },
                    },
                    {
                        config: {
                            thirdPartyId: "twitter",
                            clients: [
                                {
                                    clientId: "4398792-WXpqVXRiazdRMGNJdEZIa3RVQXc6MTpjaQ",
                                    clientSecret: "BivMbtwmcygbRLNQ0zk45yxvW246tnYnTFFq-LH39NwZMxFpdC",
                                },
                            ],
                        },
                    },
                ],
            },
        }),
        Passwordless.init({
            contactMethod: "EMAIL_OR_PHONE",
            flowType: "USER_INPUT_CODE_AND_MAGIC_LINK",
        }),
        Session.init(),
        Dashboard.init(),
        UserRoles.init(),
    ],
};
```

### Multi-factor auth

```ts
import ThirdParty from "supertokens-node/recipe/thirdparty";
import EmailPassword from "supertokens-node/recipe/emailpassword";
import Passwordless from "supertokens-node/recipe/passwordless";
import Session from "supertokens-node/recipe/session";
import { TypeInput } from "supertokens-node/types";
import Dashboard from "supertokens-node/recipe/dashboard";
import MultiFactorAuth from "supertokens-node/recipe/multifactorauth";
import AccountLinking from "supertokens-node/recipe/accountlinking";
import EmailVerification from "supertokens-node/recipe/emailverification";
import TOTP from "supertokens-node/recipe/totp";

export function getApiDomain() {
    const apiPort = process.env.VITE_APP_API_PORT || 3001;
    const apiUrl = process.env.VITE_APP_API_URL || `http://localhost:${apiPort}`;
    return apiUrl;
}

export function getWebsiteDomain() {
    const websitePort = process.env.VITE_APP_WEBSITE_PORT || 3000;
    const websiteUrl = process.env.VITE_APP_WEBSITE_URL || `http://localhost:${websitePort}`;
    return websiteUrl;
}

export const SuperTokensConfig: TypeInput = {
    supertokens: {
        // this is the location of the SuperTokens core.
        connectionURI: "https://try.supertokens.com",
    },
    appInfo: {
        appName: "SuperTokens Demo App",
        apiDomain: getApiDomain(),
        websiteDomain: getWebsiteDomain(),
    },
    // recipeList contains all the modules that you want to
    // use from SuperTokens. See the full list here: https://supertokens.com/docs/guides
    recipeList: [
        EmailPassword.init(),
        ThirdParty.init({
            signInAndUpFeature: {
                providers: [
                    // We have provided you with development keys which you can use for testing.
                    // IMPORTANT: Please replace them with your own OAuth keys for production use.
                    {
                        config: {
                            thirdPartyId: "google",
                            clients: [
                                {
                                    clientId:
                                        "1060725074195-kmeum4crr01uirfl2op9kd5acmi9jutn.apps.googleusercontent.com",
                                    clientSecret: "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",
                                },
                            ],
                        },
                    },
                    {
                        config: {
                            thirdPartyId: "github",
                            clients: [
                                {
                                    clientId: "467101b197249757c71f",
                                    clientSecret: "e97051221f4b6426e8fe8d51486396703012f5bd",
                                },
                            ],
                        },
                    },
                    {
                        config: {
                            thirdPartyId: "apple",
                            clients: [
                                {
                                    clientId: "4398792-io.supertokens.example.service",
                                    additionalConfig: {
                                        keyId: "7M48Y4RYDL",
                                        privateKey:
                                            "-----BEGIN PRIVATE KEY-----\nMIGTAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBHkwdwIBAQQgu8gXs+XYkqXD6Ala9Sf/iJXzhbwcoG5dMh1OonpdJUmgCgYIKoZIzj0DAQehRANCAASfrvlFbFCYqn3I2zeknYXLwtH30JuOKestDbSfZYxZNMqhF/OzdZFTV0zc5u5s3eN+oCWbnvl0hM+9IW0UlkdA\n-----END PRIVATE KEY-----",
                                        teamId: "YWQCXGJRJL",
                                    },
                                },
                            ],
                        },
                    },
                    {
                        config: {
                            thirdPartyId: "twitter",
                            clients: [
                                {
                                    clientId: "4398792-WXpqVXRiazdRMGNJdEZIa3RVQXc6MTpjaQ",
                                    clientSecret: "BivMbtwmcygbRLNQ0zk45yxvW246tnYnTFFq-LH39NwZMxFpdC",
                                },
                            ],
                        },
                    },
                ],
            },
        }),
        Passwordless.init({
            contactMethod: "EMAIL_OR_PHONE",
            flowType: "USER_INPUT_CODE_AND_MAGIC_LINK",
        }),
        EmailVerification.init({
            mode: "REQUIRED",
        }),
        AccountLinking.init({
            shouldDoAutomaticAccountLinking: async () => ({
                shouldAutomaticallyLink: true,
                shouldRequireVerification: true,
            }),
        }),
        MultiFactorAuth.init({
            firstFactors: ["thirdparty", "emailpassword"],
            override: {
                functions: (oI) => ({
                    ...oI,
                    getMFARequirementsForAuth: () => [
                        {
                            oneOf: [
                                MultiFactorAuth.FactorIds.TOTP,
                                MultiFactorAuth.FactorIds.OTP_EMAIL,
                                MultiFactorAuth.FactorIds.OTP_PHONE,
                            ],
                        },
                    ],
                }),
            },
        }),
        TOTP.init(),
        Session.init(),
        Dashboard.init(),
    ],
};
```

#### Multitenancy

```ts
import ThirdParty from "supertokens-node/recipe/thirdparty";
import EmailPassword from "supertokens-node/recipe/emailpassword";
import Passwordless from "supertokens-node/recipe/passwordless";
import Session from "supertokens-node/recipe/session";
import { TypeInput } from "supertokens-node/types";
import Dashboard from "supertokens-node/recipe/dashboard";
import UserRoles from "supertokens-node/recipe/userroles";

export function getApiDomain() {
    const apiPort = process.env.VITE_APP_API_PORT || 3001;
    const apiUrl = process.env.VITE_APP_API_URL || `http://localhost:${apiPort}`;
    return apiUrl;
}

export function getWebsiteDomain() {
    const websitePort = process.env.VITE_APP_WEBSITE_PORT || 3000;
    const websiteUrl = process.env.VITE_APP_WEBSITE_URL || `http://localhost:${websitePort}`;
    return websiteUrl;
}

export const SuperTokensConfig: TypeInput = {
    supertokens: {
        // this is the location of the SuperTokens core.
        connectionURI: "https://try.supertokens.com",
    },
    appInfo: {
        appName: "SuperTokens Demo App",
        apiDomain: getApiDomain(),
        websiteDomain: getWebsiteDomain(),
    },
    // recipeList contains all the modules that you want to
    // use from SuperTokens. See the full list here: https://supertokens.com/docs/guides
    recipeList: [
        ThirdParty.init(),
        EmailPassword.init(),
        Passwordless.init({
            contactMethod: "EMAIL",
            flowType: "USER_INPUT_CODE_AND_MAGIC_LINK",
        }),
        Session.init(),
        Dashboard.init(),
        UserRoles.init(),
    ],
};
```

These rules and examples apply only to [ts.ts](mdc:boilerplate/backend/shared/typescript/ts.ts) - and nowhere outside of it.




