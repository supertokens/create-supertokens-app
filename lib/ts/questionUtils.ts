import os from "os";
import { Answers, QuestionOption, RecipeQuestionOption, UserFlags } from "./types.js";

export function getPythonRunScripts(): string[] {
    if (os.platform() === "win32") {
        return [
            "pip install virtualenv",
            "python -m virtualenv venv",
            ".\\\\venv\\\\Scripts\\\\activate.bat",
            "pip install -r requirements.txt",
            "python app.py",
        ];
    }
    return [
        "pip install virtualenv",
        "virtualenv venv",
        "chmod +x venv/bin/activate",
        ". venv/bin/activate",
        "pip install -r requirements.txt",
        "python app.py",
    ];
}

export function getDjangoPythonRunScripts(): string[] {
    if (os.platform() === "win32") {
        return [
            "pip install virtualenv",
            "python -m virtualenv venv",
            ".\\\\venv\\\\Scripts\\\\activate.bat",
            "pip install -r requirements.txt",
            "python manage.py runserver",
        ];
    }
    return [
        "pip install virtualenv",
        "virtualenv venv",
        "chmod +x venv/bin/activate",
        ". venv/bin/activate",
        "pip install -r requirements.txt",
        "python manage.py runserver",
    ];
}

// Converts the options array we declare to a format inquirer can use
export function mapOptionsToChoices(options: QuestionOption[] | RecipeQuestionOption[]) {
    return options
        .filter((i) => i.shouldDisplay !== false)
        .map((option) => {
            return {
                name: option.displayName,
                value: option.value,
            };
        });
}

/**
 * This modifies the answers hash to use the correct python framework id if the user selects python
 * for their backend
 */
export function modifyAnswersForPythonFrameworks(answers: Answers) {
    let _answers = answers;

    if (answers.backend === "python") {
        _answers.backend = answers.backendPython;
    }

    return _answers;
}

export function modifyAnswersForNodeJSFrameworks(answers: Answers) {
    let _answers = answers;
    if (answers.backend === "node") {
        _answers.backend = answers.backendNodeJS;
    }
    return _answers;
}

/**
 * Decides whether the user should be prompted to select their backend.This question is skipped for full stack frameworks
 */
export function shouldSkipBackendQuestion(answers: Answers, userFlags: UserFlags): boolean {
    let frontEndInFlags = userFlags.frontend;

    if (frontEndInFlags !== undefined) {
        // Priority goes to flags
        return (
            frontEndInFlags.startsWith("next") ||
            frontEndInFlags === "capacitor" ||
            frontEndInFlags === "remix" ||
            frontEndInFlags === "astro-react" ||
            frontEndInFlags === "sveltekit" ||
            frontEndInFlags === "nuxtjs"
        );
    }

    return (
        (answers.frontend !== undefined && answers.frontend.startsWith("next")) ||
        answers.frontend === "capacitor" ||
        answers.frontend === "remix" ||
        answers.frontend === "astro-react" ||
        answers.frontend === "sveltekit" ||
        answers.frontend === "nuxtjs"
    );
}
