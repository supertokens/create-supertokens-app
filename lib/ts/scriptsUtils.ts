import { Answers, UIBuildType } from "./types.js";
import fs from "fs/promises";
import path from "path";

export async function executeSetupStepsIfExists(directoryPath: string, answers: Answers): Promise<void> {
    if (answers.ui === UIBuildType.CUSTOM && answers.frontend === "react-custom") {
        await modifyDirectoryCustomReact(directoryPath, answers);
    }
}

async function modifyDirectoryCustomReact(frontendDir: string, answers: Answers): Promise<void> {
    const recipe = answers.recipe;
    const authPath = path.join(frontendDir, "src/pages/Auth");
    const allItems = await fs.readdir(authPath);

    /**
     * Verify if the recipe folder exists
     */
    const recipeFolder = allItems.filter((item) => item !== recipe);
    if (!recipeFolder || recipeFolder.length === 0) {
        throw new Error("recipe folder not found");
    }

    /**
     * Get Recipe Dependencies which would be injected into the recipe folder
     * Step1: check if the dependency Exists
     * Step2: Remove existing folder with the dependency name from the recipe folder
     * Step3: Delete the index.tsx file from the dependency folder
     * Step4: Copy the dependency folder from the Auth folder to the recipe folder
     */
    const dependencies = getRecipeDependencies(recipe);
    for (const dependency of dependencies) {
        if (!allItems.includes(dependency)) {
            throw new Error(`Dependency ${dependency} not found`);
        }
        const dependencyPath = path.join(authPath, dependency);
        const recipeDependencyPath = path.join(authPath, recipe, dependency);
        await fs.rm(recipeDependencyPath, { recursive: true });
        await fs.rm(path.join(dependencyPath, "index.tsx"));
        await fs.cp(dependencyPath, recipeDependencyPath, { recursive: true });
    }

    /**
     * Remove all the folders except the recipe folder
     */
    for (const item of allItems) {
        if (item === recipe) {
            continue;
        }
        const itemPath = path.join(authPath, item);
        await fs.rm(itemPath, { recursive: true });
    }

    /**
     * Move all the files from the recipe folder to Auth folder
     */
    const allItemsInsideRecipe = await fs.readdir(path.join(authPath, recipe));
    for (const item of allItemsInsideRecipe) {
        const itemPath = path.join(authPath, recipe, item);
        const newPath = path.join(authPath, item);
        await fs.cp(itemPath, newPath, { recursive: true });
    }

    /**
     * Remove the recipe folder as part of cleanup
     */
    await fs.rm(path.join(authPath, recipe), { recursive: true });
}

export function getRecipeDependencies(recipe: string): string[] {
    const recipeDependencies: Record<string, string[]> = {
        thirdpartyemailpassword: ["thirdparty", "emailpassword"],
        thirdpartypasswordless: ["thirdparty", "passwordless"],
    };
    return recipeDependencies?.[recipe] || [];
}
