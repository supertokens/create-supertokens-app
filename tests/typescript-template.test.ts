import { describe, it, expect } from "vitest";
import { generateTypeScriptTemplate } from "../boilerplate/backend/shared/typescript/ts"; // Adjust path
import { UserFlags } from "../lib/ts/types"; // Adjust path

describe("TypeScript Template Generation", () => {
    it("should generate correct config for EmailPassword only (Express)", () => {
        const args = {
            configType: "emailpassword" as const,
            userArguments: { manager: "npm" } as UserFlags,
        };
        const result = generateTypeScriptTemplate(args, "express"); // Pass framework as second arg
        expect(result).toMatchSnapshot();
    });

    it("should generate correct config for ThirdParty only (Nest)", () => {
        const args = {
            configType: "thirdparty" as const,
            userArguments: { manager: "npm" } as UserFlags,
        };
        const result = generateTypeScriptTemplate(args, "nest");
        expect(result).toMatchSnapshot();
    });

    it("should generate correct config for ThirdParty + EmailPassword + OTP Email MFA (Koa)", () => {
        const args = {
            configType: "thirdpartyemailpassword" as const,
            userArguments: {
                manager: "npm" as const,
                firstfactors: ["emailpassword", "thirdparty"],
                secondfactors: ["otp-email"], // Using OTP Email for testing
            } as UserFlags,
        };
        const result = generateTypeScriptTemplate(args, "koa");
        expect(result).toMatchSnapshot();
    });

    it("should generate correct config for EmailPassword + Link Phone (Express)", () => {
        const args = {
            configType: "emailpassword" as const, // configType might be inferred differently
            userArguments: {
                manager: "npm" as const,
                firstfactors: ["emailpassword", "link-phone"],
            } as UserFlags,
        };
        const result = generateTypeScriptTemplate(args, "express");
        // Check specific parts for passwordless config
        expect(result).toContain('contactMethod: "PHONE"');
        expect(result).toContain('flowType: "MAGIC_LINK"');
        expect(result).toMatchSnapshot();
    });

    // TODO: Add more specific test cases for different factor combinations and frameworks
});
