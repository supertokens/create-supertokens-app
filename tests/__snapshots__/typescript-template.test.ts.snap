// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TypeScript Template Generation > should generate correct config for EmailPassword + Link Phone (Express) 1`] = `
"import EmailPassword from "supertokens-node/recipe/emailpassword";
import Session from "supertokens-node/recipe/session";
import Dashboard from "supertokens-node/recipe/dashboard";
import UserRoles from "supertokens-node/recipe/userroles";
import Passwordless from "supertokens-node/recipe/passwordless";
import type { TypeInput } from "supertokens-node/types";

export function getApiDomain() {
    const apiPort = 3001;
    const apiUrl = \`http://localhost:\${apiPort}\`;
    return apiUrl;
}

export function getWebsiteDomain() {
    const websitePort = 3000;
    const websiteUrl = \`http://localhost:\${websitePort}\`;
    return websiteUrl;
}

export const SuperTokensConfig: TypeInput = {
    supertokens: {
        connectionURI: "https://try.supertokens.com",
    },
    appInfo: {
        appName: "SuperTokens Demo App",
        apiDomain: getApiDomain(),
        websiteDomain: getWebsiteDomain(),
        apiBasePath: "/auth", 
        websiteBasePath: "/auth", 
    },
    recipeList: [
        EmailPassword.init(),
        Dashboard.init(),
        UserRoles.init(),
        Passwordless.init({
    contactMethod: "PHONE",
    flowType: "MAGIC_LINK"
}),
        Session.init()
    ],
};"
`;

exports[`TypeScript Template Generation > should generate correct config for EmailPassword only (Express) 1`] = `
"import EmailPassword from "supertokens-node/recipe/emailpassword";
import Session from "supertokens-node/recipe/session";
import Dashboard from "supertokens-node/recipe/dashboard";
import UserRoles from "supertokens-node/recipe/userroles";
import type { TypeInput } from "supertokens-node/types";

export function getApiDomain() {
    const apiPort = 3001;
    const apiUrl = \`http://localhost:\${apiPort}\`;
    return apiUrl;
}

export function getWebsiteDomain() {
    const websitePort = 3000;
    const websiteUrl = \`http://localhost:\${websitePort}\`;
    return websiteUrl;
}

export const SuperTokensConfig: TypeInput = {
    supertokens: {
        connectionURI: "https://try.supertokens.com",
    },
    appInfo: {
        appName: "SuperTokens Demo App",
        apiDomain: getApiDomain(),
        websiteDomain: getWebsiteDomain(),
        apiBasePath: "/auth", 
        websiteBasePath: "/auth", 
    },
    recipeList: [
        EmailPassword.init(),
        Dashboard.init(),
        UserRoles.init(),
        Session.init()
    ],
};"
`;

exports[`TypeScript Template Generation > should generate correct config for ThirdParty + EmailPassword + OTP Email MFA (Koa) 1`] = `
"import EmailPassword from "supertokens-node/recipe/emailpassword";
import ThirdParty from "supertokens-node/recipe/thirdparty";
import type { ProviderInput } from "supertokens-node/recipe/thirdparty/types";
import Session from "supertokens-node/recipe/session";
import Dashboard from "supertokens-node/recipe/dashboard";
import UserRoles from "supertokens-node/recipe/userroles";
import MultiFactorAuth from "supertokens-node/recipe/multifactorauth";
import EmailVerification from "supertokens-node/recipe/emailverification";
import AccountLinking from "supertokens-node/recipe/accountlinking";
import Passwordless from "supertokens-node/recipe/passwordless";
import type { AccountInfoWithRecipeId } from "supertokens-node/recipe/accountlinking/types";
import type { User } from "supertokens-node/types";
import type { TypeInput } from "supertokens-node/types";

export function getApiDomain() {
    const apiPort = 3001;
    const apiUrl = \`http://localhost:\${apiPort}\`;
    return apiUrl;
}

export function getWebsiteDomain() {
    const websitePort = 3000;
    const websiteUrl = \`http://localhost:\${websitePort}\`;
    return websiteUrl;
}

export const SuperTokensConfig: TypeInput = {
    supertokens: {
        connectionURI: "https://try.supertokens.com",
    },
    appInfo: {
        appName: "SuperTokens Demo App",
        apiDomain: getApiDomain(),
        websiteDomain: getWebsiteDomain(),
        apiBasePath: "/auth", 
        websiteBasePath: "/auth", 
    },
    recipeList: [
        EmailPassword.init(),
        ThirdParty.init({
            signInAndUpFeature: {
                providers: [
                    {
                        config: {
                            thirdPartyId: "google",
                            clients: [
                                {
                                    clientId: "1060725074195-kmeum4crr01uirfl2op9kd5acmi9jutn.apps.googleusercontent.com",
                                    clientSecret: "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",
                                    
                                },
                            ],
                        },
                    },
                    {
                        config: {
                            thirdPartyId: "github",
                            clients: [
                                {
                                    clientId: "467101b197249757c71f",
                                    clientSecret: "e97051221f4b6426e8fe8d51486396703012f5bd",
                                    
                                },
                            ],
                        },
                    },
                    {
                        config: {
                            thirdPartyId: "apple",
                            clients: [
                                {
                                    clientId: "4398792-io.supertokens.example.service",
                                    clientSecret: "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",
                                    additionalConfig: {
          "keyId": "7M48Y4RYDL",
          "privateKey": "-----BEGIN PRIVATE KEY-----\\nMIGTAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBHkwdwIBAQQgu8gXs+XYkqXD6Ala9Sf/iJXzhbwcoG5dMh1OonpdJUmgCgYIKoZIzj0DAQehRANCAASfrvlFbFCYqn3I2zeknYXLwtH30JuOKestDbSfZYxZNMqhF/OzdZFTV0zc5u5s3eN+oCWbnvl0hM+9IW0UlkdA\\n-----END PRIVATE KEY-----",
          "teamId": "YWQCXGJRJL"
}
                                },
                            ],
                        },
                    },
                    {
                        config: {
                            thirdPartyId: "twitter",
                            clients: [
                                {
                                    clientId: "4398792-WXpqVXRiazdRMGNJdEZIa3RVQXc6MTpjaQ",
                                    clientSecret: "BivMbtwmcygbRLNQ0zk45yxvW246tnYnTFFq-LH39NwZMxFpdC",
                                    
                                },
                            ],
                        },
                    }
                ],
            },
        }),
        Dashboard.init(),
        UserRoles.init(),
        MultiFactorAuth.init({
    firstFactors: ["emailpassword", "thirdparty"],
    override: {
        functions: (oI) => ({
            ...oI,
            getMFARequirementsForAuth: async () => [
                {
                    oneOf: [
                        MultiFactorAuth.FactorIds.OTP_EMAIL
                    ],
                },
            ],
        }),
    }
}),
        EmailVerification.init({
        mode: "REQUIRED"
    }),
        AccountLinking.init({
            shouldDoAutomaticAccountLinking: async (
                newAccountInfo: AccountInfoWithRecipeId,
                user: User | undefined,
                session: any,
                tenantId: string,
                userContext: any
            ) => {
                return {
                    shouldAutomaticallyLink: true,
                    shouldRequireVerification: true
                };
            }
        }),
        Passwordless.init({
    contactMethod: "EMAIL",
    flowType: "USER_INPUT_CODE"
}),
        Session.init()
    ],
};"
`;

exports[`TypeScript Template Generation > should generate correct config for ThirdParty only (Nest) 1`] = `
"import ThirdParty from "supertokens-node/recipe/thirdparty";
import type { ProviderInput } from "supertokens-node/recipe/thirdparty/types";
import Session from "supertokens-node/recipe/session";
import Dashboard from "supertokens-node/recipe/dashboard";
import UserRoles from "supertokens-node/recipe/userroles";
import type { TypeInput } from "supertokens-node/types";

export function getApiDomain() {
    const apiPort = 3001;
    const apiUrl = \`http://localhost:\${apiPort}\`;
    return apiUrl;
}

export function getWebsiteDomain() {
    const websitePort = 3000;
    const websiteUrl = \`http://localhost:\${websitePort}\`;
    return websiteUrl;
}

export const SuperTokensConfig: TypeInput = {
    supertokens: {
        connectionURI: "https://try.supertokens.com",
    },
    appInfo: {
        appName: "SuperTokens Demo App",
        apiDomain: getApiDomain(),
        websiteDomain: getWebsiteDomain(),
        apiBasePath: "/auth", 
        websiteBasePath: "/auth", 
    },
    recipeList: [
        ThirdParty.init({
            signInAndUpFeature: {
                providers: [
                    {
                        config: {
                            thirdPartyId: "google",
                            clients: [
                                {
                                    clientId: "1060725074195-kmeum4crr01uirfl2op9kd5acmi9jutn.apps.googleusercontent.com",
                                    clientSecret: "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",
                                    
                                },
                            ],
                        },
                    },
                    {
                        config: {
                            thirdPartyId: "github",
                            clients: [
                                {
                                    clientId: "467101b197249757c71f",
                                    clientSecret: "e97051221f4b6426e8fe8d51486396703012f5bd",
                                    
                                },
                            ],
                        },
                    },
                    {
                        config: {
                            thirdPartyId: "apple",
                            clients: [
                                {
                                    clientId: "4398792-io.supertokens.example.service",
                                    clientSecret: "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",
                                    additionalConfig: {
          "keyId": "7M48Y4RYDL",
          "privateKey": "-----BEGIN PRIVATE KEY-----\\nMIGTAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBHkwdwIBAQQgu8gXs+XYkqXD6Ala9Sf/iJXzhbwcoG5dMh1OonpdJUmgCgYIKoZIzj0DAQehRANCAASfrvlFbFCYqn3I2zeknYXLwtH30JuOKestDbSfZYxZNMqhF/OzdZFTV0zc5u5s3eN+oCWbnvl0hM+9IW0UlkdA\\n-----END PRIVATE KEY-----",
          "teamId": "YWQCXGJRJL"
}
                                },
                            ],
                        },
                    },
                    {
                        config: {
                            thirdPartyId: "twitter",
                            clients: [
                                {
                                    clientId: "4398792-WXpqVXRiazdRMGNJdEZIa3RVQXc6MTpjaQ",
                                    clientSecret: "BivMbtwmcygbRLNQ0zk45yxvW246tnYnTFFq-LH39NwZMxFpdC",
                                    
                                },
                            ],
                        },
                    }
                ],
            },
        }),
        Dashboard.init(),
        UserRoles.init(),
        Session.init()
    ],
};"
`;
